File: haskell/flood.hs
{- |
Module      :  flood-algorithm.hs
Description :  This is the flood algorithm written in Haskell 
Copyright   :  (c) Author: Philippe Lumpkin
License     :  Open Source License
Maintainer  :  phiglump@ut.utm.edu
Stability   :  unstable | <experimental> | provisional | stable | frozen
Discussed high level code with Nicholas Pica and Hunter Haislip
-}

import Data.List
-- ^ imported for nub and sort 

reachable :: [(Char, Char)] -> Char -> [Char]
reachable graph start = sort ( nub (reachable' graph start))
-- ^ This is the main function that will take the list and the starting point
-- It will also sort and remove duplicates from the list 

reachable' :: [(Char, Char)] -> Char -> [Char]
reachable' graph start = start: restCompare [start] graph
-- ^ this function essentially just adds the start node to a list so that restCompare will accept it
-- still functionally works the same 

restCompare :: [Char] -> [(Char, Char)] -> [Char]
restCompare [] graph = []
restCompare (n:ns) graph = compareStart ++ restCompare compareStart removeTuple ++ restCompare ns removeTuple
    where 
    compareStart = (compare' n graph)
    removeTuple = (remove n graph)
-- ^ The middle part of the concatenation from my understanding seems to be more for positioning and for the recursion
-- the first will get the first set of neighbors from the neighbors list and the last part is sending the rest of the list
-- thru the restcompare again. 

compare' :: Char -> [(Char, Char)] -> [Char]
compare' start [] = []
compare' start (x:xs)
    | start == fst x = snd x: compare' start xs
    | otherwise = compare' start xs
-- ^ comparing the starting node to the head of all the tuples in the given list

remove :: Char -> [(Char, Char)] -> [(Char, Char)]
remove start [] = []
remove start (x:xs)
    | start == fst x = remove start xs
    | otherwise = x: remove start xs 
-- ^ This is my remove function that updates the graph after the compare function has already gathered what it needed to know
-- It works very similar to how compare' worked but instead of printing the list of neighbors it instead removes the head on a match condition

-- [('a', 'f'), ('a', 'h'), ('a', 'c'), ('c', 'd'), ('f', 'd')]
-- [('k','f'),('x','l'),('n','s'),('b','a'),('f','q'),('a','c'),('f','s'),('r','h'),('d','t'),('b','y'),('h','g'),('r','y'),('c','c'),('a','g'),('v','j'),('v','l'),('s','u'),('i','j'),('y','e'),('y','v'),('a','f'),('f','g'),('z','h'),('y','y'),('g','q'),('y','s'),('b','t'),('l','j'),('m','q'),('p','k'),('p','b'),('t','w'),('k','j'),('q','c')]
-- ^ test cases that i used as reference, the second one uses loops like c to c and like c to b and b to c.

File: java/Flood.java
/**
 *File: Flood.java
 *This is the completed build for the flood algorithm
 *It will take the info provided in the file and pass it into a Hashmap
 *From there it is parsed to find all of the reachable nodes from the given starting node
 *And lastly it will have any duplicates removed and be sorted and then printed. 
 *@author Philippe Lumpkin
 *@version 2.1
 *@since 2020-04-03
 */

import java.util.*;
import java.io.File;
import java.io.FileNotFoundException;

public class Flood {
    /**
     *This simply will just remove the Dups and Sort the list before printing for the user to see. 
     *@param neighbors the list of all of the confirmed reachables
     **/
    public void rmDupsAndSort(List<String> neighbors) {
	List<String> temp = new ArrayList<String>();

	for(String a : neighbors) {
	    if (!temp.contains(a)) {
		temp.add(a);
	    }
	}
	Collections.sort(temp);
	System.out.println(temp);
    }
    /**
     *This function is the parser that goes thru the list by first getting the first set of reachable nodes from 
     *the very start. Essentially jump starting the loop afterwords.
     *It will then make sure that what it will try to check isn't null and also that it isn't in the already 
     *checked nodes, otherwise it would probably end up with loops or redundant checks/info. 
     *If it has passed the if statements it will begin to gather all of the reachable's reachables,
     *and add to the reachable list as it goes until it is finished.
     *@param m this is the map of the edges
     *@param start this is the starting node
     **/
    public void parser(Map<String, List<String>> m, String start) {
	List<String> neighbors = new ArrayList<String>();
	List<String> temp = new ArrayList<String>();
	String checked = "";
	if (m.get(start) != null) {
	    neighbors = m.get(start);
	    
	    for (int i = 0; i < neighbors.size(); i++) {
		if (m.get(neighbors.get(i)) != null) {
		    if (checked.indexOf(neighbors.get(i)) == -1){
			temp = m.get(neighbors.get(i));
			neighbors.addAll(temp);
			checked = checked + neighbors.get(i);
		    }
		}
	    }
	    Flood floodObject3 = new Flood();
	    floodObject3.rmDupsAndSort(neighbors);
	}
	else {
	    System.out.println("The typed in start doesn't exist within in the list.");
	}
    }
    /**
     *This is the method that takes all the input from the given graph file
     *and store all of it into a hash map. This seemed the most appealing to me as it represents the edges conviently
     *@param args the arguments passed into the prgram during execution
     **/
    public void getData(String[] args) {
	try {
	    Scanner inFile = new Scanner(new File(args[0]));
	    String start = args[1];
	    Map<String, List<String>> m = new HashMap<String, List<String>>();
	    while(inFile.hasNext()) {
		String a = inFile.next();
		String b = inFile.next();
		m.putIfAbsent(a, new ArrayList<String>());
		m.get(a).add(b);
	    }
	    inFile.close();
	    Flood floodObject2 = new Flood();
	    floodObject2.parser(m, start);
	}
	catch (FileNotFoundException e) {
	    System.out.println("Error: File " + args[0] + " doesn't exist.");
	    e.printStackTrace();
	}
    }
    
    /**
     *This is the main, or the entry point into the program and it will create an object and call the get data function to begin
     *@param args The single parameter will be the graph 
     */
    public static void main(String[] args) {
	Flood floodObject = new Flood();
	floodObject.getData(args);
    }
}
